PhysicsTools/Heppy/python/physicsutils/JetReCalibrator.py

```diff
@@ -40,15 +57,20 @@ def correctAll(self,jets,rho,delta=0,metShift=[0,0]):
             print "Warning: %d out of %d jets flagged bad by JEC." % (len(badJets), len(jets))
         for bj in badJets:
             jets.remove(bj)
+        for j in jets:
+            for sepcorr in self.separateJetCorrectors.keys():
+                setattr(j,"CorrFactor_"+sepcorr,self.getCorrection(j,rho,0,[0,0],corrector=self.separateJetCorrectors[sepcorr]))
 
-    def getCorrection(self,jet,rho,delta=0,metShift=[0,0]):
+    def getCorrection(self,jet,rho,delta=0,metShift=[0,0],corrector=None):
         """Calculates the correction factor of a jet without modifying it
```

Comment:
It is not true that this function does not modify the jet. If delta!=0, the attribute "jetEnergyCorrUncertainty" of the jet is set. The additional JEC levels can only be called with delta==0 to avoid overwriting jet attributes calculated with the standard corrector.